### PR TITLE
Runtime tetragon

### PR DESCRIPTION
**Pull request by pxp928**
Signed-off-by: pxp928 <parth.psu@gmail.com>

Fixes Issue #212 - Adding runtime visibility and enforcement into frsca

Future updates will need to sign and attest the runtime output based on the Pipelines/taskruns being run. This would utilize witness. 

```
The first way is to observe the raw json output from the stdout container log:

kubectl logs -n kube-system -l app.kubernetes.io/name=tetragon -c export-stdout -f
```
REQUIREMENT:

The base kernel should support [BTF](https://github.com/cilium/tetragon#btf-requirement) or the BTF file should be placed where Tetragon can read it. Tetragon repository provides a [Vagrantfile](https://github.com/cilium/tetragon/blob/main/Vagrantfile) that can be use to install a vagrant box for running Tetragon with BTF requirement.

The testing of this was done on GKE. Other cloud provider kubernetes offering would also work.

### CODE REVIEW

**Comment by rgreinho** [comment]
Is it supposed to do something right now?

I ran `make quickstart`, and I do see it being deployed and starting (see snippet bellow), but after that I am not sure how to check that it does what it is supposed to do. 

```bash
Setting up Tetragon...
+ helm upgrade --install tetragon /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/chart --values /Users/rgreinhofer/projects/buildsec/ssf/platform/components/tetragon/values.yaml --namespace kube-system --wait
Release "tetragon" does not exist. Installing it now.
NAME: tetragon
LAST DEPLOYED: Wed Jul  6 12:21:20 2022
NAMESPACE: kube-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
+ kubectl rollout status -n kube-system daemonset/tetragon
Waiting for daemon set "tetragon" rollout to finish: 0 of 1 updated pods are available...
daemon set "tetragon" successfully rolled out
+ echo -e '\033[32mEnable tracing policies for Tetragon...\033[0m'
Enable tracing policies for Tetragon...
+ kubectl apply -f /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/tracingpolicy/v0.8.0/sys_write_follow_fd_prefix.yaml
tracingpolicy.cilium.io/sys-read-follow-prefix created
+ kubectl apply -f /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/tracingpolicy/v0.8.0/tcp-connect.yaml
tracingpolicy.cilium.io/connect created
```

The log command you provided does not return anything.

```bash
╭ ssf on  pr/249 via 🐹 v1.18.3 ⌛ 8s
╰❯ kubectl logs -n kube-system ds/tetragon -c export-stdout


```

**Comment by pxp928** [comment]
> Is it supposed to do something right now?
> 
> I ran `make quickstart`, and I do see it being deployed and starting (see snippet bellow), but after that I am not sure how to check that it does what it is supposed to do.
> 
> ```shell
> Setting up Tetragon...
> + helm upgrade --install tetragon /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/chart --values /Users/rgreinhofer/projects/buildsec/ssf/platform/components/tetragon/values.yaml --namespace kube-system --wait
> Release "tetragon" does not exist. Installing it now.
> NAME: tetragon
> LAST DEPLOYED: Wed Jul  6 12:21:20 2022
> NAMESPACE: kube-system
> STATUS: deployed
> REVISION: 1
> TEST SUITE: None
> + kubectl rollout status -n kube-system daemonset/tetragon
> Waiting for daemon set "tetragon" rollout to finish: 0 of 1 updated pods are available...
> daemon set "tetragon" successfully rolled out
> + echo -e '\033[32mEnable tracing policies for Tetragon...\033[0m'
> Enable tracing policies for Tetragon...
> + kubectl apply -f /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/tracingpolicy/v0.8.0/sys_write_follow_fd_prefix.yaml
> tracingpolicy.cilium.io/sys-read-follow-prefix created
> + kubectl apply -f /Users/rgreinhofer/projects/buildsec/ssf/platform/vendor/tetragon/tracingpolicy/v0.8.0/tcp-connect.yaml
> tracingpolicy.cilium.io/connect created
> ```
> 
> The log command you provided does not return anything.
> 
> ```shell
> ╭ ssf on  pr/249 via 🐹 v1.18.3 ⌛ 8s
> ╰❯ kubectl logs -n kube-system ds/tetragon -c export-stdout
> ```

try this command. I updated the above to match 
`kubectl logs -n kube-system -l app.kubernetes.io/name=tetragon -c export-stdout -f`

**Comment by rgreinho** [comment]
Same thing: 

```bash
╭ ssf on  pr/249 via 🐹 v1.18.3
╰❯ kubectl logs -n kube-system -l app.kubernetes.io/name=tetragon -c export-stdout

╭ ssf on  pr/249 via 🐹 v1.18.3
╰❯
```

Even if I run another example in another tab, I do not see any logs.

**Comment by rgreinho** [comment]
eBPF won't work on a Mac, so we will need someone with a linux host to test it.

**Comment by pxp928** [comment]
REQUIREMENT:

The base kernel should support [BTF](https://github.com/cilium/tetragon#btf-requirement) or the BTF file should be placed where Tetragon can read it. Tetragon repository provides a [Vagrantfile](https://github.com/cilium/tetragon/blob/main/Vagrantfile) that can be use to install a vagrant box for running Tetragon with BTF requirement.

The testing of this was done on GKE. Other cloud provider kubernetes offering would also work.

**Comment by sudo-bmitch** [comment]
So it runs in my environment and I definitely get logs. What I'm looking for is either a test of something it prevents from running, or a malicious build that it detects. Just looking at the container stdout, I don't think anyone could sort out if there's something wrong from that, it's far too much of a firehose of raw logs.

**Comment by pxp928** [comment]
> So it runs in my environment and I definitely get logs. What I'm looking for is either a test of something it prevents from running, or a malicious build that it detects. Just looking at the container stdout, I don't think anyone could sort out if there's something wrong from that, it's far too much of a firehose of raw logs.

I agree. I can add in a test in there that shows it either blocking or detecting malicious builds

**Comment by mlieberman85** [comment]
@pxp928 We can look at my old "rust" compiler build where a malicious compiler injects a binary at runtime into memory and runs it. https://github.com/mlieberman85/supply-chain-examples/tree/main/build_with_ebpf

**Comment by pxp928** [comment]
> @pxp928 We can look at my old "rust" compiler build where a malicious compiler injects a binary at runtime into memory and runs it. https://github.com/mlieberman85/supply-chain-examples/tree/main/build_with_ebpf

Yup that's what I was thinking. Will do.

**Comment by rgreinho** [comment]
Should we had a check not to deploy Tetragon if the platform is not linux? Just asking because the pod keeps on crashing in the background and there is no specific reason for me to have it running since I am using a mac M1.

**Comment by pxp928** [comment]
hmmm. For some reason the new pipeline I created is hanging. When I tested it on GKE. It completes very quickly. @sudo-bmitch can you try running the `make example-trufflehog` locally on your machine.

File: platform/vendor/vendor.yaml [hunk 1]
```diff
@@ -63,3 +63,4 @@ files:
     destination_dir: "cert-manager/release"
     sha256: "517b57b60e37288ced462bb541f3f5dabb53f59f40c262387403862985b01ae8"
     validation_type: "sha256"
+
```
[the remark this answers is not in the document]
Comment (pxp928) [reply platform/vendor/vendor.yaml]: fixed.

**Comment by sudo-bmitch** [comment]
@pxp928 I'm getting an ImagePullBackoff:

```
Back-off pulling image "madhuakula/k8s-goat-build-code"
```

File: platform/vendor/tetragon/tracingpolicy/v0.8.0/disallow-trufflehog.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+apiVersion: cilium.io/v1alpha1
+kind: TracingPolicy
+metadata:
+  name: "disallow-trufflehog"
+spec:
+  kprobes:
+  - call: "fd_install"
```
Comment (mlieberman85) [inline platform/vendor/tetragon/tracingpolicy/v0.8.0/disallow-trufflehog.yaml:7]: Can we change this to an exec instead?

**Pull request closed without merging**